### PR TITLE
perf(ggml-cuda): tensor-core head-256 flash attention on RDNA4

### PR DESCRIPTION
**Pull request by dusterbloom**
## What

Qwen3.5/3.6/3.8 dense-hybrid targets attend at head_dim=256, and the RDNA4 dispatch capped the MMA fattn path at head 128, so prefill fell to the generic tile kernel. This PR lifts the cap and routes head-256 prefill through tensor cores by default on RDNA4 (`DFLASH27B_FA256_MMA=0` opts out): the raw-MMA kernel everywhere, plus the rocWMMA kernel below a crossover when built with `GGML_HIP_ROCWMMA_FATTN` (off by default). The crossover is `DFLASH27B_FA256_WMMA_MAX_KV` (default 32768), named and tunable, with its measured A/B data in the fattn.cu comment.

## Validated results: end-to-end prefill, Qwen3.8-27B UD-IQ4_XS

Cold prefill, q8_0 KV, chunk-512, per-prompt `--max-ctx` (blog methodology), measured on one Radeon AI PRO R9700 (gfx1201, ROCm 7.2), single runs (no repeats; the shared box limits variance control). Before = tile kernel (the previous dispatch), after = this PR's default.

| Prompt tokens | Before tok/s | After tok/s | Speedup |
|---:|---:|---:|---:|
| 8.7K | 899 | 918 | +2% |
| 12.2K | 899 | 947 | +5% |
| 15.5K | 841 | 911 | +8% |
| 23.9K | 785 | 829 | +6% |
| 30.2K | 715 | 817 | +14% |
| 34.0K | 691 | 785 | +14% |
| 43.7K | 616 | 729 | +18% |
| 59.5K | 525 | 640 | +22% |
| 65.3K | 486 | 608 | +25% |
| 120.4K | 343 | 458 | +34% |

Gain grows monotonically with context because attention is a rising share of prefill wall time (tile-kernel attention was the slope). Output text parity confirmed against the tile kernel.

Kernel-level attention (q8_0 KV, nq=512): 1.71–1.96x the tile kernel from 8K to 131K KV.

## Qualification

- `test_fattn_mma256`: differential vs ggml CPU flash_attn_ext, route pinned by launch counters. Enforced tolerances 1e-3 (f16 KV) / 2e-3 (q8_0 KV); measured max diffs 5e-4. Matrix: S up to 65536 (both sides of the WMMA/MMA crossover), a ragged-S fallback assertion, attention-sink and logit-softcap cases, and a kill-switch mode (`DFLASH27B_FA256_MMA=0`) that verifies the tile fallback route + correctness.
- `test_rocwmma_layout`: rocWMMA fragment-layout probes for RDNA4 (all PASS).
- `bench_fattn_mma256`: throughput bench (not in the unit-test list).

## Caveats

- Full-model eval suites (NIAH etc.) were not run as part of this PR; the kernel differential + text parity are the gates. Env opt-outs remain for A/B.
- The rocWMMA f16-accumulator mma chain compiles to zero results on gfx1201 (fragment semantics probe clean in isolation); the D=256 arms accumulate KQ in f32, gated to RDNA4 so MUSA/RDNA3/CDNA flag builds keep their historical f16 accumulation.
- The combine-tile barrier hunk (`np > 1 || DV/2 > nbatch_combine`) fixes a real tile-overwrite race for single-warp-per-column Turing configs (DV=256, nbatch_combine=64); on RDNA4 the np>=2 disjunct already fires, so it is inert there. Kept in this PR because the hunk is one line; happy to split if reviewers prefer.
- The flag stays off by default: unqualified on RDNA3 and non-ROCm platforms.

### CODE REVIEW

File: server/CMakeLists.txt [hunk 1]
```diff
@@ -1033,6 +1054,52 @@ if(DFLASH27B_TESTS)
...
+        set_target_properties(test_rocwmma_layout PROPERTIES HIP_ARCHITECTURES "${_dflash_archs}")
+        target_compile_definitions(test_rocwmma_layout PRIVATE GGML_USE_HIP)
+        find_path(DFLASH27B_ROCWMMA_INCLUDE_DIR rocwmma/rocwmma.hpp
+            HINTS "${_dflash_rocm_root}/include" /opt/rocm/include
+            NO_DEFAULT_PATH)
+        if(DFLASH27B_ROCWMMA_INCLUDE_DIR)
+            target_include_directories(test_rocwmma_layout PRIVATE ${DFLASH27B_ROCWMMA_INCLUDE_DIR})
+        endif()
+        target_link_libraries(test_rocwmma_layout PRIVATE ggml ggml-base ${DFLASH27B_GGML_BACKEND_TARGET} hip::host)
+        list(APPEND _raw_unit_test_targets test_rocwmma_layout)
```
[the remark this answers is not in the document]
Comment (dusterbloom) [reply server/CMakeLists.txt]: Fixed in `12808b2d`/`72575edf` (verified on gfx1201).

File: server/deps/llama.cpp/ggml/src/ggml-cuda/fattn.cu [hunk 2]
```diff
@@ -4456,7 +4479,10 @@ static best_fattn_kernel ggml_cuda_get_best_fattn_kernel(const int device, const
     }
 
     // Use the WMMA kernel if possible:
-    if (ggml_cuda_should_use_wmma_fattn(cc) && K->ne[1] % FATTN_KQ_STRIDE == 0 && Q->ne[0] != 40 && Q->ne[0] != 72 && Q->ne[0] != 512 && Q->ne[0] != 576) {
+    // On RDNA4 the rocWMMA kernel is not qualified (fragment layouts do not
+    // match the hand-rolled softmax reductions), so it is reachable only
+    // through the env-gated head-256 block below.
+    if (ggml_cuda_should_use_wmma_fattn(cc) && !GGML_CUDA_CC_IS_RDNA4(cc) && K->ne[1] % FATTN_KQ_STRIDE == 0 && Q->ne[0] != 40 && Q->ne[0] != 72 && Q->ne[0] != 256 && Q->ne[0] != 512 && Q->ne[0] != 576) {
```
[the remark this answers is not in the document]
Comment (dusterbloom) [reply server/deps/llama.cpp/ggml/src/ggml-cuda/fattn.cu]: Fixed in `12808b2d`/`72575edf` (verified on gfx1201).

File: server/deps/llama.cpp/ggml/src/ggml-cuda/fattn-mma-f16.cuh [hunk 3]
```diff
@@ -1531,7 +1536,11 @@ static __device__ __forceinline__ void flash_attn_ext_f16_process_tile(
                 }
             }
         }
-        if (np > 1) {
+        // With DV/2 > nbatch_combine the next iteration overwrites the
+        // combine tile while warps from this one may still be reading it,
+        // even when np == 1. Only DV <= 128 (single iteration) can skip
```
[the remark this answers is not in the document]
Comment (dusterbloom) [reply server/deps/llama.cpp/ggml/src/ggml-cuda/fattn-mma-f16.cuh]: Fixed in `12808b2d`/`72575edf` (verified on gfx1201).

File: server/test/test_rocwmma_layout.cpp [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+        half_t *d_A, *d_B, *d_C;
+        float  *d_Cf;
+        HIP_CHECK(hipMalloc(&d_A, Ah.size() * sizeof(half_t)));
+        HIP_CHECK(hipMalloc(&d_B, Bh.size() * sizeof(half_t)));
+        HIP_CHECK(hipMalloc(&d_C, (size_t) v.ldm_c * N * sizeof(half_t)));
+        HIP_CHECK(hipMalloc(&d_Cf, (size_t) v.ldm_c * N * sizeof(float)));
+        HIP_CHECK(hipMemcpy(d_A, Ah.data(), Ah.size() * sizeof(half_t), hipMemcpyHostToDevice));
+        HIP_CHECK(hipMemcpy(d_B, Bh.data(), Bh.size() * sizeof(half_t), hipMemcpyHostToDevice));
+
+        if (v.ldm_a == 256) {
```
[the remark this answers is not in the document]
Comment (dusterbloom) [reply server/test/test_rocwmma_layout.cpp]: Fixed in `12808b2d`/`72575edf` (verified on gfx1201).

File: server/test/test_fattn_mma256.cpp [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+    ggml_tensor * maskf = ggml_new_tensor_2d(cctx, GGML_TYPE_F16, c.S, c.nq);
+    ggml_tensor * ref = ggml_flash_attn_ext(cctx, Qr, Kr, Vr, maskf, scale, 0.0f, 0.0f);
+    ggml_set_output(ref);
+
+    ggml_cgraph * cgraph = ggml_new_graph(cctx);
+    ggml_build_forward_expand(cgraph, ref);
+
+    ggml_gallocr_t galloc_g = ggml_gallocr_new(ggml_backend_get_default_buffer_type(gpu));
+    ggml_gallocr_t galloc_c = ggml_gallocr_new(ggml_backend_get_default_buffer_type(cpu));
+    if (!galloc_g || !galloc_c ||
```
[the remark this answers is not in the document]
Comment (dusterbloom) [reply server/test/test_fattn_mma256.cpp:139]: Fixed in `12808b2d`/`72575edf` (verified on gfx1201).

File: server/test/test_fattn_mma256.cpp [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+    if (hipGetDeviceProperties(&props, 0) != hipSuccess ||
+        std::strncmp(props.gcnArchName, "gfx12", 5) != 0) {
+        std::printf("[fattn-mma256] SKIP: requires gfx12 (RDNA4)\n");
+        return 77;
+    }
+    // Must precede the first GPU fattn dispatch: fattn.cu reads the opt-in
+    // once per process on the first dispatch call. An explicit value in the
+    // environment wins, so DFLASH27B_FA256_MMA=0 can force the tile kernel
+    // for route A/B diagnosis.
+    if (getenv("DFLASH27B_FA256_MMA") == nullptr &&
```
[the remark this answers is not in the document]
Comment (dusterbloom) [reply server/test/test_fattn_mma256.cpp:271]: Fixed in `12808b2d`/`72575edf` (verified on gfx1201).

File: server/test/test_fattn_mma256.cpp [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+        setenv("DFLASH27B_FA256_MMA", "1", 1) != 0) {
```
[the remark this answers is not in the document]
Comment (dusterbloom) [reply server/test/test_fattn_mma256.cpp:272]: Fixed in `12808b2d`/`72575edf` (verified on gfx1201).

File: server/deps/llama.cpp/ggml/src/ggml-cuda/fattn.cu [hunk 8]
```diff
@@ -4488,6 +4514,46 @@ static best_fattn_kernel ggml_cuda_get_best_fattn_kernel(const int device, const
...
+        while (gqa_ratio % (2*gqa_ratio_eff) == 0 && gqa_ratio_eff < ncols2_max) {
+            gqa_ratio_eff *= 2;
+        }
+        // DFLASH27B_FA256_WMMA forces the rocWMMA kernel on every head-256
+        // RDNA4 shape (A/B debugging); DFLASH27B_FA256_MMA only upgrades
+        // prefill-sized batches. Default policy (both unset): prefill-sized
+        // head-256 batches take the rocWMMA kernel when the build has it,
+        // falling back to the raw-MMA kernel at very long KV where its
+        // stream-K tiling wins (measured crossover ~32K on gfx1201).
+        if (fa256_wmma && ggml_cuda_should_use_wmma_fattn(cc) && K->ne[1] % FATTN_KQ_STRIDE == 0) {
```
[the remark this answers is not in the document]
Comment (dusterbloom) [reply server/deps/llama.cpp/ggml/src/ggml-cuda/fattn.cu:4546]: Not a bug: `launch_fattn` (fattn-common.cuh:1009) converts quantized K/V to an F16 pool buffer before launching when `need_f16_K/V` is set, and the WMMA case function passes `true, true`. The q8_0 qualification cases (test_fattn_mma256, S=8192/16384) pass at 5e-4 through the WMMA route, which is the empirical proof of the conversion path.

File: server/docs/ENVIRONMENT.md [hunk 9]
```diff
@@ -17,6 +17,8 @@ consolidation of this list into CLI flags is tracked as follow-up work.
 | Variable | Default | Purpose |
 |---|---|---|
 | `DFLASH_ADAPTIVE_SPEC_WIDTH` | unset | BURN-IN: =1 enables the shared acceptance-feedback verify-width controller. Fixed width is the production default; backend-specific overrides take precedence. |
+| `DFLASH27B_FA256_MMA` | 1 on RDNA4 | KILL SWITCH (burn-in): =0 restores the generic tile kernel for head-256 attention on RDNA4. With the default 1, prefill-sized head-256 batches take the rocWMMA kernel below ~32K KV and the raw-MMA kernel beyond (rocWMMA requires a `GGML_HIP_ROCWMMA_FATTN` build; without it the raw-MMA kernel covers all shapes). |
+| `DFLASH27B_FA256_WMMA` | unset | A/B: =1 forces the rocWMMA kernel on every head-256 RDNA4 shape, bypassing the KV-length crossover (requires a `GGML_HIP_ROCWMMA_FATTN` build). |
```
[the remark this answers is not in the document]
Comment (dusterbloom) [reply server/docs/ENVIRONMENT.md]: Fixed in `3ac7...` - the force branch also requires `K->ne[1] % FATTN_KQ_STRIDE == 0`, so the doc now says exactly that.

[the remark this answers is not in the document]
Comment (dusterbloom) [reply server/docs/ENVIRONMENT.md]: Fixed in 5942e343: the doc now reads "head-256 RDNA4 shapes whose KV length is a multiple of 256". (Correction of the hash typo in the previous reply.)

File: server/test/bench_mmq_prefill.cpp [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    const auto t0 = std::chrono::steady_clock::now();
+    const int iters = 5;
+    for (int i = 0; i < iters; ++i) {
+        ggml_backend_graph_compute(gpu, gf);
+    }
+    hipDeviceSynchronize();
+    const auto t1 = std::chrono::steady_clock::now();
+
+    const double ms = std::chrono::duration<double, std::milli>(t1 - t0).count() / iters;
+    const double gbs = (double) ggml_nbytes(w) / ms / 1e6; // GB / ms -> TB/s-ish
```
[the remark this answers is not in the document]
Comment (dusterbloom) [reply server/test/bench_mmq_prefill.cpp]: Fixed on this branch in 757c1023 (and on the phase-2 branch in 547ad428): the field now prints GB/s, and the bench was reworked to measure the MMQ path with F32 activations plus an F16-activation variant for the cublas-converted fallback.

File: server/deps/llama.cpp/ggml/src/ggml-cuda/fattn.cu [hunk 11]
```diff
@@ -4488,6 +4514,57 @@ static best_fattn_kernel ggml_cuda_get_best_fattn_kernel(const int device, const
...
+        return e && atoi(e) != 0;
+    }();
+    // KV length above which the raw-MMA kernel takes over from rocWMMA in
+    // flag builds. Measured on gfx1201 (q8_0 KV, nq=512): rocWMMA wins at
+    // 8K (4.24 vs 5.32 ms) and 16K (8.55 vs 10.17), raw-MMA wins at 32K
+    // (19.45 vs 19.77), 64K (37.87 vs 39.77) and 131K (74.80 vs 79.00).
+    // The crossover lies in the unmeasured 16-32K band; override with
+    // DFLASH27B_FA256_WMMA_MAX_KV for A/B.
+    static const int64_t fa256_wmma_max_kv = [] {
+        const char * e = getenv("DFLASH27B_FA256_WMMA_MAX_KV");
```
[the remark this answers is not in the document]
Comment (dusterbloom) [reply server/deps/llama.cpp/ggml/src/ggml-cuda/fattn.cu]: Fixed in f404f3e8: DFLASH27B_FA256_WMMA_MAX_KV is now in the documented-variable table (default 32768, crossover semantics) and the flat inventory.

File: server/deps/llama.cpp/ggml/src/ggml-cuda/fattn-wmma-f16.cu [hunk 12]
```diff
@@ -606,9 +614,13 @@ void ggml_cuda_flash_attn_ext_wmma_f16(ggml_backend_cuda_context & ctx, ggml_ten
...
                     ggml_cuda_flash_attn_ext_wmma_f16_case<128, cols_per_block, float>(ctx, dst);
                     break;
-                // case 256:
-                //     ggml_cuda_flash_attn_ext_wmma_f16_case<256, cols_per_block, float>(ctx, dst);
-                //     break;
+                case 256:
+                    if (rdna4) {
+                        ggml_cuda_flash_attn_ext_wmma_f16_case<256, cols_per_block, float>(ctx, dst);
+                    } else {
+                        ggml_cuda_flash_attn_ext_wmma_f16_case<256, cols_per_block, half>(ctx, dst);
```
[the remark this answers is not in the document]
Comment (dusterbloom) [reply server/deps/llama.cpp/ggml/src/ggml-cuda/fattn-wmma-f16.cu]: Valid - fixed in f404f3e8: the prec=F32 cols32 256 arm now uses the float accumulator unconditionally, matching every other head size in that switch. The RDNA4 f16-acc gate applies only to the default-precision arms.

File: server/test/test_fattn_mma256.cpp [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+    std::vector<Case> cases = {
+        {   512,  512, false, expect_tc, false, 0.0f,  tol_mul*1e-3f },
+        {  1024,   64, false, expect_tc, false, 0.0f,  tol_mul*1e-3f },
+        {  4096,   64, false, expect_tc, false, 0.0f,  tol_mul*1e-3f },
+        {  8192,   64, true,  expect_tc, false, 0.0f,  tol_mul*2e-3f },
+        { 16384,   64, true,  expect_tc, false, 0.0f,  tol_mul*2e-3f },
+        { 65536,   64, true,  expect_tc, false, 0.0f,  tol_mul*2e-3f },
+        {  3000,   64, false, false,      false, 0.0f,  tol_mul*2e-3f },
+        {  4096,   64, false, expect_tc, true,  0.0f,  tol_mul*1e-3f },
+        {  4096,   64, false, expect_tc, false, 50.0f, tol_mul*1e-3f },
```
[the remark this answers is not in the document]
Comment (dusterbloom) [reply server/test/test_fattn_mma256.cpp:304]: Not a bug: the WMMA kernel does implement logit softcap (the softmax carries a use_logit_softcap instantiation, tanh in the float branch, h2-exp in the half branch). Empirically: the flag-build test matrix runs the S=4096 softcap case through the WMMA route (launch counter pinned, below the 32K crossover) and passes at 5e-4 vs the CPU reference.
